dbcrossbar/src/cmd/cp.rs

```diff
@@ -107,8 +112,38 @@ pub(crate) async fn run(ctx: Context, opt: Opt) -> Result<()> {
         // certain degree of parallelism. This is where all the actual work happens,
         // and this what controls how many "input driver" -> "output driver"
         // connections are running at any given time.
-        result_stream.consume_with_parallelism(4).await?;
+        result_stream.consume_with_parallelism(4).await?
+    };
+
+    // Optionally display `dests`, depending on a combination of
+    // `--display-output-locators` and the defaults for `to_locator`.
+    let display_output_locators = to_locator.display_output_locators();
+    match (opt.display_output_locators, display_output_locators) {
+        // The user passed `--display-output-locators`, but displaying them is
+        // forbidden (probably because we wrote actual data to standard output).
+        (true, DisplayOutputLocators::Never) => Err(format_err!(
+            "cannot use --display-output-locators with {}",
+            to_locator
+        )),
+
+        // We want to display our actual output locators.
+        (true, _) | (false, DisplayOutputLocators::ByDefault) => {
+            for dest in dests {
+                let dest_str = dest.to_string();
+                if dest_str.contains('\n') || dest_str.contains('\r') {
+                    // If we write out this locator, it would be split between
+                    // lines, causing an ambiguity for any parsing program.
+                    return Err(format_err!(
+                        "cannot output locator with newline: {:?}",
+                        dest_str
+                    ));
+                }
+                println!("{}", dest);
```

Comment:
can this block? is there an async version? or do we not care here?

Reply:
Excellent question.

My thinking:

1. `tokio` actually has a multithreaded executor, so short blocks are _probably_ OK, except for things like `wait` that deadlock instead of block.
2. At this point, we've finished doing all our main processing.
3. This is a really tiny amount of output, just a list of locators.

But I think I need to tear this out and actually generate a `BoxStream<BoxLocator>`, and then use:

 ```rust
let stdout_sink = FramedWrite::new(io::stdout(), LinesCodec::new())`
dests_stream.forward(stdout_sink).compat().await?;
```

This would print out the destinations as we go, instead of doing it all in one batch at the end.

Reply:
I guess the block will usually be negligible

I'm trying to think if there is a case where it would NOT be negligible; perhaps if the pipe consumer can't keep up with the producer; but I guess then that isn't really a recoverable situation anyway; i wonder what the max pipe buffer size is; it appears to be quite small by default according to `man 7 pipe`; 65k; but also tune-able

Reply:
if it did happen though I guess the rust process would silently stop forever; since it would never hit the epoll() call in it's event loop again; the only way around it would be to use signals but even then signal handlers can't do much and who wants to mess with signals anyway

Reply:
but theoretically you could create a rust async timer that periodically and increments a counter; then you can set up a periodic timer SIGNAL the handler for which ensures that the counter increments periodically; after some timeout it could terminate the process with an error. I guess it just depends on what the desired behavior is. but really it's probably best to just never do sync IO inside a async code.

Reply:
Yeah, the only thing we're printing here is a list of locators:

```
bigml:dataset/1
bigml:dataset/2
bigml:dataset/3
bigml:dataset/4
...
```

If you actually have 65k of locators, this could block. But even if it blocks, it can't deadlock, because `println!` never touches `tokio`. So what happens is that the executor thread will stall until the output buffer drains below 65k of locators. But we're using a multithreaded executor.

So basically, this is bad form but it should never break in practice. The _real_ problem is that it waits until _all_ output locators are known, and then it prints them out all at once. We want to print them out as we go, so that `bigml-parallel` can consume them incrementally.

To do this, I need to tweak the async stream a bit. You're welcome to pair with me if you're interested.